sources/lib/api/gcp/data/__init__.py

```diff
@@ -27,10 +27,109 @@ def sql(sql_template, **kwargs):
     sql_template: the template of the SQL statement with named placeholders.
     **kwargs: the dictionary of name/value pairs to use for placeholder values.
   Returns:
-    The formatted SQL statement with placeholders replaced with their values, and
-    an array of Javascript UDFs referenced by the code.
+    The formatted SQL statement with placeholders replaced with their values.
   Raises:
     Exception if a placeholder was found in the SQL statement, but did not have
     a corresponding argument value.
   """
   return _util.Sql.format(sql_template, kwargs)
+
+
+def tokenize(sql):
+  """ This is a crude tokenizer for our limited purposes.
```

Comment:
Side note -- this shouldn't be public or perhaps even in __init__.py.

Secondly coroutines would be simpler to read through rather than hand-coded state machines, assuming they are an option. See http://eli.thegreenplace.net/2009/08/29/co-routines-as-an-alternative-to-state-machines/

I guess we're going to invariably end up in sql parsing. We should think about how we do this correctly across the scenarios -- parsing out UDF references, parsing begin/end of queries, comments in %%sql modules etc.